Db/is event monitor support windows

derekwbrown opened the pull request:
### What does this PR do?

This PR lays the groundwork for future event monitor support for Windows.
The implementation is currently disabled.

### Motivation

The motivation is to chart a path forward for sharing code between Linux & Windows fo rthe
event monitor support.  

### Additional Notes

The feature is disabled in the PR; this is simply laying the groundwork for a future feature.

### Possible Drawbacks / Trade-offs

### Describe how to test/QA your changes

N/A

### Reviewer's Checklist

- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [x] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [x] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

ogaca-dd commented on pkg/windowsdriver/include/procmonapi.h:
```diff
@@ -0,0 +1,54 @@
+#pragma once
```
`pkg/windowsdriver` doesn't seem to have a code owner.

clarkb7 commented on pkg/security/resolvers/process/resolver_windows.go:
```diff
@@ -0,0 +1,77 @@
+
+	e.Process.PIDContext.Pid = uint32(e.Pid)
+	e.Process.Argv0 = file
+	e.Process.Argv = strings.Split(commandLine, " ")
```
I'm concerned about splitting the command line on space here without at least taking quoting rules into account.

[process-agent](https://github.com/DataDog/datadog-agent/blob/4bd973b1fec5ef621a86d9125c775880dda200ea/pkg/process/procutil/process_windows.go#L575-L618) has some custom parsing rules

Windows provides [CommandLineToArgv](https://learn.microsoft.com/en-us/windows/win32/api/shellapi/nf-shellapi-commandlinetoargvw). But there's some discussion about the use of this method [here](https://github.com/golang/go/issues/15588).

Since the Windows CreateProcess APIs take a commandline and leave it up to the application to parse it as it will, some apps (like cmd.exe) have custom parsing rules, so we may not be able to have a one sized fits all split/unsplit for args.

derekwbrown replied:
Point is well taken.
However, point of this PR is to get something in so we're not always chasing merge conflicts, etc.
Note that the PR doesn't include the driver, so there's nothing feeding this at this point.  This is more foundational.

I can certainly add a large comment here which indicates exactly how naive this implementation is, so it could be enhanced as we add functionality.

richl-dd replied:
At a minimum I think we need the comment; however, if there are straightforward / preferred alternatives (like the one Branden linked from process-agent) that improve the foundation from the start, I think we should consider using them to mitigate the risk of "we'll come back to this later" but not actually doing so.

Is there a reason we can't / shouldn't use the code from process-agent that Branden linked? That seems like it'd provide consistency throughout the agent for how the Windows command lines will be parsed and eventually represented in the UI. It'd also make it easy for process-agent to compare / contrast this implementation with their own if/when they shift to this implementation.

brycekahle reviewed (approved):
shallow review of ebpf-platform owned files

paulcacheux commented on pkg/security/probe/opts_windows.go:
Not sure why you need this while it seems to be the same file as `opts_linux.go` ?

paulcacheux reviewed (approved):
Small nit, but LGTM. Thanks !

derekwbrown commented on tasks/winbuildscripts/dobuild.bat:
```diff
@@ -40,8 +40,8 @@ if "%TARGET_ARCH%" == "x86" (
 
-if not exist \dev\go\src\github.com\DataDog\datadog-agent exit /b 100
-cd \dev\go\src\github.com\DataDog\datadog-agent || exit /b 101
+rem if not exist \dev\go\src\github.com\DataDog\datadog-agent exit /b 100
+rem cd \dev\go\src\github.com\DataDog\datadog-agent || exit /b 101
```
fixed.  Was accidental checkin

lrbhatia commented:
This PR does not have a description or any other fields filled out.

derekwbrown committed:
resolve conflicts

add codeowners

fix my mistakes in resolving conflicts on linux side

Add license header to omnibus download file

remove unneeded entries.

Try to fix copyright linting

fix copyright regex

Fix copyright merge again (sort of)
fix rebase mistake on linux side

iglendd commented on pkg/security/resolvers/process/resolver_windows.go:
```diff
@@ -0,0 +1,77 @@
+	p.maplock.Lock()
+	defer p.maplock.Unlock()
+	if _, ok := p.processes[pid]; ok {
+		delete(p.processes, pid)
```
Just curios about overall plan dealing with PID collision, there are many pitfalls, e.g. removal PID notification thread reached user mode after notification of creating new process with the same pid. Or something needed entry for a pid at the time and reference for "older" process with the same pid.

derekwbrown replied:
would definitely be worth following up offline.
But, the current strategy I've used elsewhere is pid in combination with process start time.

bkabrda commented on omnibus/config/software/datadog-windows-procmon-driver.rb:
```diff
@@ -0,0 +1,21 @@
+name "datadog-windows-procmon-driver"
```
Let's add the notice header here:
```
# Unless explicitly stated otherwise all files in this repository are licensed
# under the Apache License Version 2.0.
# This product includes software developed at Datadog (https:#www.datadoghq.com/).
# Copyright 2016-present Datadog, Inc.
```

richl-dd replied:
Added, thanks for noticing this!

bkabrda reviewed (approved):
LGTM, but please add the notice header to the new file before merging.

clarkb7 commented:
bypassing, agent-shared-components not a codeowner
